triedb/pathdb/buffer.go

```diff
@@ -0,0 +1,150 @@
+}
+
+// newBuffer initializes the buffer with the provided states and trie nodes.
+func newBuffer(limit int, nodes *nodeSet, layers uint64) *buffer {
+	// Don't panic for lazy users if any provided set is nil
+	if nodes == nil {
+		nodes = newNodeSet(nil)
+	}
+	return &buffer{
+		layers: layers,
+		limit:  uint64(limit),
+		nodes:  nodes,
+	}
+}
+
+// node retrieves the trie node with node path and its trie identifier.
+func (b *buffer) node(owner common.Hash, path []byte) (*trienode.Node, bool) {
+	return b.nodes.node(owner, path)
+}
+
+// commit merges the provided states and trie nodes into the buffer.
+//
+// This operation does not take ownership of the passed maps, which belong to
+// the bottom-most diff layer. Instead, it holds references to the given maps,
+// which are safe to copy.
+func (b *buffer) commit(nodes *nodeSet) *buffer {
+	b.layers++
+	b.nodes.merge(nodes)
+	return b
+}
+
+// revert is the reverse operation of commit. It also merges the provided states
+// and trie nodes into the buffer. The key difference is that the provided state
+// set should reverse the changes made by the most recent state transition.
+func (b *buffer) revert(db ethdb.KeyValueReader, nodes map[common.Hash]map[string]*trienode.Node) error {
+	// Short circuit if no embedded state transition to revert
+	if b.layers == 0 {
+		return errStateUnrecoverable
+	}
+	b.layers--
+
+	// Reset the entire buffer if only a single transition left
+	if b.layers == 0 {
+		b.reset()
+		return nil
+	}
+	b.nodes.revert(db, nodes)
+	return nil
+}
+
+// reset cleans up the disk cache.
+func (b *buffer) reset() {
+	b.layers = 0
+	b.nodes.reset()
+}
+
+// empty returns an indicator if buffer is empty.
+func (b *buffer) empty() bool {
+	return b.layers == 0
+}
+
+// full returns an indicator if the size of accumulated content exceeds the
+// configured threshold.
+func (b *buffer) full() bool {
+	return b.size() > b.limit
+}
+
+// size returns the approximate memory size of the held content.
+func (b *buffer) size() uint64 {
+	return b.nodes.size
+}
+
+// allocBatch returns a database batch with pre-allocated buffer.
+func (b *buffer) allocBatch(db ethdb.KeyValueStore) ethdb.Batch {
+	return db.NewBatchWithSize(b.nodes.dbsize() * 11 / 10) // extra 10% for potential pebble internal stuff
+}
+
+// flush persists the in-memory dirty trie node into the disk if the configured
+// memory threshold is reached. Note, all data must be written atomically.
+func (b *buffer) flush(db ethdb.KeyValueStore, freezer ethdb.AncientWriter, nodesCache *fastcache.Cache, id uint64) error {
+	// Ensure the target state id is aligned with the internal counter.
+	head := rawdb.ReadPersistentStateID(db)
+	if head+b.layers != id {
+		return fmt.Errorf("buffer layers (%d) cannot be applied on top of persisted state id (%d) to reach requested state id (%d)", b.layers, head, id)
+	}
+	// Terminate the state snapshot generation if it's active
+	var (
+		start = time.Now()
+		batch = b.allocBatch(db)
```

Comment:
wdyt about inlining this? The semantics of the allocBatch method are a bit weird to me, you're passing a db and creating a batch from that with the internal dbsize. 
Since this is the only spot its used, I would just inline it